File: core/transcription/src/commonMain/kotlin/dev/audiocompanion/transcription/SonioxContext.kt
package dev.audiocompanion.transcription

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import kotlinx.serialization.json.putJsonArray

/** Soniox transcription `context` object (real-time config + async REST body). */
@Serializable
data class SonioxContext(
    val text: String? = null,
    val terms: List<String>? = null,
)

/** Builds Soniox `context` JSON for WebSocket config when text or terms are present. */
internal fun buildSonioxContextJsonObject(
    contextText: String?,
    contextTerms: List<String>,
): JsonObject? {
    val text = contextText?.trim()?.takeIf { it.isNotEmpty() }
    val terms = contextTerms.filter { it.isNotBlank() }.distinct()
    if (text == null && terms.isEmpty()) return null
    return buildJsonObject {
        text?.let { put("text", it) }
        if (terms.isNotEmpty()) {
            putJsonArray("terms") { terms.forEach { add(JsonPrimitive(it)) } }
        }
    }
}

/** Maps lambdas into a [SonioxContext] for async REST requests. */
internal fun sonioxContextFrom(
    contextText: () -> String?,
    contextTerms: () -> List<String>,
): SonioxContext? {
    val text = contextText()?.trim()?.takeIf { it.isNotEmpty() }
    val terms = contextTerms().filter { it.isNotBlank() }.distinct()
    if (text == null && terms.isEmpty()) return null
    return SonioxContext(
        text = text,
        terms = terms.takeIf { it.isNotEmpty() },
    )
}

File: ios/PebbleAudioKit/Sources/Transcription/Providers/SonioxContext.swift
import Foundation

// Port of `core/transcription/.../SonioxContext.kt`.

/// Soniox transcription `context` object (real-time config + async REST body).
public struct SonioxContext: Codable, Sendable, Equatable {
    public var text: String?
    public var terms: [String]?

    public init(text: String? = nil, terms: [String]? = nil) {
        self.text = text
        self.terms = terms
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encodeIfPresent(text, forKey: .text)
        try container.encodeIfPresent(terms, forKey: .terms)
    }
}

/// Builds Soniox `context` JSON for WebSocket config when text or terms are present.
func buildSonioxContextJsonObject(
    contextText: String?,
    contextTerms: [String]
) -> [String: Any]? {
    guard let context = sonioxContextFrom(contextText: { contextText }, contextTerms: { contextTerms })
    else { return nil }
    var object: [String: Any] = [:]
    if let text = context.text { object["text"] = text }
    if let terms = context.terms { object["terms"] = terms }
    return object
}

/// Maps closures into a `SonioxContext` for async REST requests (nil when there is no context).
func sonioxContextFrom(
    contextText: @Sendable () -> String?,
    contextTerms: @Sendable () -> [String]
) -> SonioxContext? {
    let text = contextText()?.trimmingCharacters(in: .whitespacesAndNewlines).ifEmptyNil
    var seen = Set<String>()
    let terms = contextTerms().filter { term in
        let isBlank = term.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
        return !isBlank && seen.insert(term).inserted
    }
    if text == nil && terms.isEmpty { return nil }
    return SonioxContext(text: text, terms: terms.isEmpty ? nil : terms)
}

extension String {
    fileprivate var ifEmptyNil: String? { isEmpty ? nil : self }
}
